cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaBrokerConfigurationBuilder.java

```diff
@@ -70,37 +71,64 @@ public class KafkaBrokerConfigurationBuilder {
+    private final NodeRef node;
 
     /**
      * Broker configuration template constructor
      *
      * @param reconciliation    The reconciliation
      * @param brokerId          The ID of the broker
-     * @param useKRaft          Indicates whether KRaft is used or not
+     * @param useKRaft          Indicates whether KRaft feature gate is enabled or not
+     * @param node              NodeRef instance
+     * @param kafkaMetadataConfigState  Represents the state of the Kafka metadata configuration
      */
-    public KafkaBrokerConfigurationBuilder(Reconciliation reconciliation, String brokerId, boolean useKRaft) {
+    public KafkaBrokerConfigurationBuilder(Reconciliation reconciliation, String brokerId, boolean useKRaft, NodeRef node, KafkaMetadataStateManager.KafkaMetadataConfigurationState kafkaMetadataConfigState) {
         printHeader();
         this.reconciliation = reconciliation;
         this.brokerId = brokerId;
         this.useKRaft = useKRaft;
+        this.node = node;
+        this.kafkaMetadataConfigState = kafkaMetadataConfigState;
 
         // Render the node/broker ID into the config file
         configureNodeOrBrokerId();
     }
 
+    /**
+     * @return if the Kafka metadata configuration state is from ZooKeeper-based up to KRaft migration (and dual-write) going on
+     */
+    private boolean isZooKeeperOrMigration() {
+        return this.kafkaMetadataConfigState.ordinal() <= KafkaMetadataStateManager.KafkaMetadataConfigurationState.MIGRATION.ordinal();
+    }
+
+    /**
+     * @return if the Kafka metadata configuration state is from KRaft post-migration up to the full KRaft-based
+     */
+    private boolean isPostMigrationOrKRaft() {
+        return this.kafkaMetadataConfigState.ordinal() >= KafkaMetadataStateManager.KafkaMetadataConfigurationState.POST_MIGRATION.ordinal();
+    }
+
+    /**
+     * @return if KRaft is going to be configured depending on the node role and the Kafka metadata configuration state
+     */
+    private boolean isKRaftInConfiguration() {
+        return (node.controller() && this.kafkaMetadataConfigState.ordinal() >= KafkaMetadataStateManager.KafkaMetadataConfigurationState.PRE_MIGRATION.ordinal()) ||
+                (node.broker() && this.kafkaMetadataConfigState.ordinal() >= KafkaMetadataStateManager.KafkaMetadataConfigurationState.MIGRATION.ordinal());
+    }
+
     /**
      * Renders the broker.id or node.id configurations
      */
     private void configureNodeOrBrokerId()   {
         printSectionHeader("Node / Broker ID");
 
-        if (useKRaft) {
-            writer.println("node.id=" + brokerId);
-        } else {
+        // Node ID is ignored when not using Kraft mode => but it defaults to broker ID when not set.
+        // We set it here in the configuration explicitly to avoid never ending rolling updates.
+        writer.println("node.id=" + brokerId);
```

Comment:
In the previous code, the `node.id` was set in both Zk and KRaft clusters and this is going to happen with this code as well, so nothing changed actually. The current `if` is now adding the `broker.id` not only if cluster is Zk but also during migration (because brokers are still connected to Zk and this field is needed).

Reply:
You are right. I misread the code change.